browser/ui/views/location_bar/brave_shields_page_info_controller.cc

```diff
@@ -0,0 +1,95 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/location_bar/brave_shields_page_info_controller.h"
+
+#include "base/check_deref.h"
+#include "brave/browser/ui/views/page_info/brave_page_info_bubble_view.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "chrome/browser/ui/views/location_bar/location_icon_view.h"
+#include "chrome/browser/ui/views/page_info/page_info_bubble_specification.h"
+#include "chrome/browser/ui/views/page_info/page_info_bubble_view.h"
+#include "content/public/browser/navigation_controller.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/views/view_utils.h"
+
+BraveShieldsPageInfoController::BraveShieldsPageInfoController(
+    TabStripModel* tab_strip_model,
+    LocationIconView* location_icon_view)
+    : tab_strip_model_(CHECK_DEREF(tab_strip_model)),
+      location_icon_view_(CHECK_DEREF(location_icon_view)) {
+  tab_strip_model_->AddObserver(this);
+  if (auto* shields_helper = GetShieldsHelper()) {
+    shields_helper->AddObserver(this);
+  }
+}
+
+BraveShieldsPageInfoController::~BraveShieldsPageInfoController() {
+  if (auto* shields_helper = GetShieldsHelper()) {
+    shields_helper->RemoveObserver(this);
+  }
+}
+
+void BraveShieldsPageInfoController::OnTabStripModelChanged(
+    TabStripModel* tab_strip_model,
+    const TabStripModelChange& change,
+    const TabStripSelectionChange& selection) {
+  if (selection.active_tab_changed()) {
+    if (auto* shields_helper = GetShieldsHelper(selection.old_contents)) {
+      shields_helper->RemoveObserver(this);
+    }
+    if (auto* shields_helper = GetShieldsHelper(selection.new_contents)) {
+      shields_helper->AddObserver(this);
+    }
```

Comment:
I think this part could make unexpected bug because `GetShieldsHelper()` will give helper if passed arg is null.
`selection.old_contents` or `selection.new_contents` could be null then `GetShieldsHelper()` will give helper for active tab always. is it ok?